crypto/chacha_drbg.h

```diff
@@ -21,23 +21,36 @@
 #define __CHACHA_DRBG__
 
 #include "chacha20poly1305/chacha20poly1305.h"
+#include "sha2.h"
 
-// Very fast deterministic random bit generator inspired by CTR_DRBG in NIST SP
-// 800-90A
+// A very fast deterministic random bit generator based on CTR_DRBG in NIST SP
+// 800-90A. Chacha is used instead of a block cipher in the counter mode, SHA256
+// is used as a derivation function. The highest supported security strength is
+// at least 256 bits. Reseeding is left up to caller.
 
-#define CHACHA_DRBG_KEY_LENGTH 16
+#define CHACHA_DRBG_KEY_LENGTH 32
+#define CHACHA_DRBG_COUNTER_LENGTH 8
 #define CHACHA_DRBG_IV_LENGTH 8
-#define CHACHA_DRBG_SEED_LENGTH (CHACHA_DRBG_KEY_LENGTH + CHACHA_DRBG_IV_LENGTH)
+#define CHACHA_DRBG_SEED_LENGTH \
+  (CHACHA_DRBG_KEY_LENGTH + CHACHA_DRBG_COUNTER_LENGTH + CHACHA_DRBG_IV_LENGTH)
+
+#define CHACHA_DRBG_DERIVATION_FUNCTION_PREFIX_LENGTH (1 + 4)
+// 1 = sizeof(counter), 4 = sizeof(output_length) in derivation_function
+#define CHACHA_DRBG_DERIVATION_FUNCTION_PADDING 9
+// padding in SHA256
+#define CHACHA_DRBG_DERIVATION_FUNCTION_BLOCK_LENGTH SHA256_BLOCK_LENGTH
```

Comment:
This is true but only because entropy that goes to `derivation_function` always fits _one_ block of SHA256. Since this is something I wanted to be easy to change, I chose more universal but cumbersome solution.

Reply:
There is nothing universal or easy about doing computations with these internal constants in `core/embed/unix/drbg.h` and `core/embed/trezorhal/drbg.h`. It obfuscates the meaning of the `_Static_assert`s in the `core/embed/*/drbg.h` files, it violates the DRY principle and the principle of decoupling.

Reply:
> There is nothing universal

I think you didn't get the point. Once we decide (maybe even in this PR) that `drbg_init` uses "two blocks" of entropy, the assert becomes:

```
_Static_assert(CHACHA_DRBG_DERIVATION_FUNCTION_BLOCK_LENGTH -
                       2 * CHACHA_DRBG_DERIVATION_FUNCTION_PREFIX_LENGTH -
                       CHACHA_DRBG_DERIVATION_FUNCTION_PADDING ==
                   DRBG_INIT_TRNG_ENTROPY_LENGTH,
               "");
```

If you insist on decoupling principle what about something like:
```
#define CHACHA_DRBG_OPTIMAL_RESEED_LENGTH(block_count)             \
  (block_count * (CHACHA_DRBG_DERIVATION_FUNCTION_BLOCK_LENGTH -   \
                  CHACHA_DRBG_DERIVATION_FUNCTION_PREFIX_LENGTH) - \
   CHACHA_DRBG_DERIVATION_FUNCTION_PADDING)
```
and

```
_Static_assert(CHACHA_DRBG_OPTIMAL_RESEED_LENGTH(1) ==
                   DRBG_INIT_TRNG_ENTROPY_LENGTH,
               "");
```

Reply:
> Once we decide (maybe even in this PR) that `drbg_init` uses "two blocks" of entropy, the assert becomes:
> 
> ```
> _Static_assert(CHACHA_DRBG_DERIVATION_FUNCTION_BLOCK_LENGTH -
>                        2 * CHACHA_DRBG_DERIVATION_FUNCTION_PREFIX_LENGTH -
>                        CHACHA_DRBG_DERIVATION_FUNCTION_PADDING ==
>                    DRBG_INIT_TRNG_ENTROPY_LENGTH,
>                "");
> ```

That makes no sense. As far as I can tell, the expression above is wrong, which perfectly proves my point about it being difficult to use and the meaning being obfuscated.

> If you insist on decoupling principle what about something like:
> 
> ```
> #define CHACHA_DRBG_OPTIMAL_RESEED_LENGTH(block_count)             \
>   (block_count * (CHACHA_DRBG_DERIVATION_FUNCTION_BLOCK_LENGTH -   \
>                   CHACHA_DRBG_DERIVATION_FUNCTION_PREFIX_LENGTH) - \
>    CHACHA_DRBG_DERIVATION_FUNCTION_PADDING)
> ```
> 
> and
> 
> ```
> _Static_assert(CHACHA_DRBG_OPTIMAL_RESEED_LENGTH(1) ==
>                    DRBG_INIT_TRNG_ENTROPY_LENGTH,
>                "");
> ```

Yes, that's what I am saying all along, except that the expression is wrong here too, although differently this time. It would be much easier to read if it were written as
```c
#define CHACHA_DRBG_OPTIMAL_RESEED_LENGTH(block_count) ((block_count) * SHA256_BLOCK_LENGTH - 1 - 4 - 1 - 8)
```
and the magic numbers were explained in a comment. The names of the constants are so long and it's so difficult to read that it's no surprise you got it wrong twice in a row and differently each time. Just imagine how difficult it would be for someone else who didn't write the code and is trying to make sense of it and make changes in it. That's precisely what I meant when I was saying that it's obfuscated and there is nothing easy or universal about it. Calling it `CHACHA_DRBG_OPTIMAL_RESEED_LENGTH` makes it easy to use and makes the meaning clearer and parametrizing it with `block_count` is what will make it "universal".

On a side note, I am not really convinced this should even be checked by the static asserts, but I also have nothing against it. So fine by me as long as it's written properly.